File: fastLPR_R/R/fastlpr_create.R
# Copyright (c) 2024-2025 Ying Wang, Min Li
# SPDX-License-Identifier: GPL-3.0-or-later

#' Mirrors: fastLPR/utility/core/ fastLPR_create.m
#'
#' Initializes the regression structure used by the fastLPR pipeline.
#' Mirrors the MATLAB helper line-by-line, relying on the exact utility
#' functions defined in fastlpr_exact_utils.R and existing helpers such as
#' `set_defaults` and `multispace`.
#'
#' @keywords internal
fastlpr_create <- function(x, y, h = NULL, opt = NULL) {
  if (is.null(opt)) opt <- list()
  if (is.null(h)) h <- matrix(numeric(0), nrow = 0)

  # Process complex-valued predictors (split columns with imaginary part)
  x_is_complex <- fastlpr_is_complex(x, 1)
  if (is.matrix(x)) {
    real_part <- Re(x)
  } else {
    real_part <- matrix(Re(x), ncol = 1)
  }
  if (any(x_is_complex)) {
    imag_cols <- Im(x)[, x_is_complex, drop = FALSE]
    x <- cbind(real_part, imag_cols)
  } else {
    x <- real_part
  }

  Tx <- nrow(x)
  dx <- ncol(x)

  if (length(dim(y)) == 2L) {
    Ty <- nrow(y)
    dy <- ncol(y)
    opt$y_corr_bandwidth <- FALSE
  } else if (length(dim(y)) == 3L) {
    dims_y <- dim(y)
    Ty <- dims_y[1]
    dhy <- dims_y[2]
    dy <- dims_y[3]
    opt$y_corr_bandwidth <- TRUE
  } else {
    stop("fastlpr_create: y must be a matrix or 3D array.")
  }

  if (Tx != Ty) {
    stop(
      sprintf(
        "fastlpr_create: x (%d rows) and y (%d rows) must match.",
        Tx,
        Ty
      )
    )
  }

  # Default options (mirroring MATLAB set_defaults calls)
  opt <- set_defaults(opt, "y_type_out", "mean")
  opt <- set_defaults(opt, "kernel_type", "gaussian")
  opt <- set_defaults(opt, "order", 0)
  opt <- set_defaults(opt, "Nratio", 1)
  default_N <- ceiling(opt$Nratio * (Tx^(1 / dx)))
  opt <- set_defaults(opt, "N", default_N)
  N <- opt$N
  if (length(N) == 1L) {
    N <- rep(N, dx)
  }
  opt$N <- N
  opt$Nratio <- N[1] / (Tx^(1 / dx))

  # FFT padding: use power-of-2 length for speed and accuracy
  # Must match MATLAB default (flag_power2=true) for cross-language consistency
  # Note: For 3D, power-of-2 causes L³ to explode (e.g., N=50 -> L=128 -> 15x more memory)
  # but is required for correct accuracy. Use flag_power2=FALSE manually if memory-constrained.
  opt <- set_defaults(opt, "flag_power2", TRUE)
  opt <- set_defaults(opt, "accuracy", max(6 - ceiling(log10(Tx)), 4))
  opt <- set_defaults(opt, "nufft_deconv", opt$accuracy > 6)
  opt <- set_defaults(opt, "dstd", 0)
  opt <- set_defaults(opt, "calc_dof", TRUE)
  opt <- set_defaults(opt, "num_dof_sample", 10)

  if (!is.list(opt$y_grid_opt)) {
    opt$y_grid_opt <- list(
      # Changed from "spline" to "linear" to match:
      # 1. MATLAB's griddedInterpolant('linear') used in cv_fastKDE
      # 2. The hardcoded "linear" in fastkde_lcv() for LCV computation
      # This ensures consistency between LCV values and fpp.evaluate()
      Method = "linear",
      ExtrapolationMethod = "linear"
    )
  }

  opt <- set_defaults(opt, "compact", TRUE)
  opt <- set_defaults(opt, "y_corr_bandwidth", FALSE)
  opt <- set_defaults(opt, "y_gcv_metric", NULL)
  opt <- set_defaults(opt, "y_grid_method", "griddedInterpolant")
  opt <- set_defaults(opt, "isgpu", FALSE)

  dtype <- typeof(y)

  converted <- fastlpr_bit_convert(opt$accuracy, x, y)
  x <- converted[[1]]
  y <- converted[[2]]
  dtype <- typeof(y)

  ndcolon <- replicate(dx, quote(expr = ), simplify = FALSE)

  if (isTRUE(opt$flag_power2)) {
    L <- 2^ceiling(log2(2 * N - 1))
  } else {
    L <- N + 2
  }
  if (length(L) == 1L) {
    L <- rep(L, dx)
  }

  # Ensure x is a matrix before saving to xraw
  if (!is.matrix(x)) {
    x <- as.matrix(x)
  }

  xraw <- x
  zs <- zscore(x)
  x <- zs$z

  # Ensure x is a matrix (zscore should return matrix, but safeguard)
  if (!is.matrix(x)) {
    x <- as.matrix(x)
  }

  x_min <- apply(x, 2, min)
  x_max <- apply(x, 2, max)
  x_scale <- x_max - x_min
  x_delta <- x_scale / N

  x <- sweep(
    x,
    2,
    (x_max + x_min) / 2,
    FUN = "-"
  )
  x_min <- apply(x, 2, min)
  x_max <- apply(x, 2, max)
  x_mean <- colMeans(x)

  qin <- matrix(0, nrow = 2, ncol = dx)
  pad_left <- floor((L - N) / 2)
  qin[1, ] <- pad_left + 1
  qin[2, ] <- pad_left + N

  # MATLAB passes z-scored x to fastLPR_kdf (see fastLPR_create.m:228)
  # Grid is created from z-scored data for kernel computation
  kdf_result <- fastlpr_kdf(x, h, N, opt)
  kdf <- kdf_result$kdf
  ihbad <- kdf_result$ihbad
  dh <- kdf_result$dh
  dh_input <- kdf_result$dh_input
  h <- kdf_result$h
  lwp <- kdf_result$lwp

  if (opt$y_corr_bandwidth && exists("dhy") && dh != dhy && dhy != 1) {
    stop(
      sprintf(
        "fastlpr_create: bandwidth count (%d) must match responses (%d)",
        dh,
        dhy
      )
    )
  }

  if (dx == 1) {
    idx_knot <- seq.int(qin[1], qin[2])
    idx_range <- seq(-0.5, 0.5 - 1 / L, length.out = L)
    hf <- 1 / L

    if (N != L) {
      knot_scale <- idx_range[idx_knot[length(idx_knot)]] - idx_range[idx_knot[1]]
      knot <- (x - x_min) / x_scale * knot_scale + idx_range[idx_knot[1]]
    } else {
      knot_scale <- idx_range[length(idx_range)] - idx_range[1]
      knot <- (x - x_min) / x_scale * knot_scale - 0.5
    }
  } else {
    idx_knot <- vector("list", dx)
    idx_range <- vector("list", dx)
    hf <- numeric(dx)
    knot <- matrix(0, nrow = Tx, ncol = dx)
    knot_scale <- numeric(dx)

    for (d in seq_len(dx)) {
      idx_knot[[d]] <- seq.int(qin[1, d], qin[2, d])
      idx_range[[d]] <- seq(-0.5, 0.5 - 1 / L[d], length.out = L[d])
      hf[d] <- 1 / L[d]

      if (N[d] != L[d]) {
        knot_scale[d] <- idx_range[[d]][idx_knot[[d]][length(idx_knot[[d]])]] -
          idx_range[[d]][idx_knot[[d]][1]]
        knot[, d] <- (x[, d] - x_min[d]) / x_scale[d] * knot_scale[d] +
          idx_range[[d]][idx_knot[[d]][1]]
      } else {
        knot_scale[d] <- idx_range[[d]][length(idx_range[[d]])] - idx_range[[d]][1]
        knot[, d] <- (x[, d] - x_min[d]) / x_scale[d] * knot_scale[d] - 0.5
      }
    }
  }

  grid_params <- fastlpr_compute_grid_params(Ty, opt$accuracy)
  Msp <- grid_params$Msp

  if (dx == 1) {
    qout_val <- (L - N) / 2
    qout <- c(ceiling(qout_val) + 1, ceiling(qout_val) + N)
  } else {
    # NOTE (2025-11-28): Use rbind (not cbind) to match MATLAB format
    # MATLAB qout format: row 1 = start indices, row 2 = end indices
    # Each column corresponds to one dimension
    qout <- rbind(
      ceiling((L - N) / 2) + 1,  # Row 1: start indices for all dims
      ceiling((L - N) / 2) + N   # Row 2: end indices for all dims
    )
  }

  xlist <- multispace(
    apply(xraw, 2, min),
    apply(xraw, 2, max),
    N
  )

  if (!is.list(xlist)) {
    xlist <- lapply(seq_len(dx), function(d) xlist[d, ])
  }
  xlist <- lapply(xlist, as.numeric)

  regs <- list(
    xraw = xraw,
    x = x,
    y = y,
    Tx = Tx,
    Ty = Ty,
    dx = dx,
    dy = dy,
    dtype = dtype,
    Msp = Msp,
    L = L,
    N = N,
    ndcolon = ndcolon,
    x_min = x_min,
    x_max = x_max,
    x_scale = x_scale,
    x_mean = x_mean,
    x_std = zs$sigma,
    x_delta = x_delta,
    h = h,
    kdf = kdf,
    ihbad = ihbad,
    dh = dh,
    dh_input = dh_input,
    qin = qin,
    qout = qout,
    knot_scale = if (dx == 1) knot_scale else knot_scale,
    knot = knot,
    hf = hf,
    xlist = xlist,
    opt = opt,
    y_isreal = !fastlpr_is_complex(y, "all"),
    lwp = lwp
  )

  regs
}

File: fastLPR/utility/core/fastlpr_create.m
function [regs]=fastlpr_create(x,y,h,opt)
% FASTLPR_CREATE - Initialize regression structure for fast local polynomial regression
%
% Creates and initializes the regression structure containing all parameters,
% grids, and precomputed values needed for fast LPR with NUFFT acceleration.
%
% Syntax:
%   regs = fastlpr_create(x, y)
%   regs = fastlpr_create(x, y, h)
%   regs = fastlpr_create(x, y, h, opt)
%
% Inputs:
%   x   - Predictor data (Tx x dx matrix, can be complex-valued)
%   y   - Response data (Ty x dy matrix or Ty x dh x dy array)
%   h   - Bandwidth candidates (dh x dx matrix, optional)
%   opt - Options structure (optional, see below)
%
% Output:
%   regs - Regression structure containing:
%          Data: .x, .y, .xraw, .Tx, .Ty, .dx, .dy
%          Grid: .N, .L, .xlist, .knot, .qin, .qout
%          Kernel: .h, .kdf, .dh, .lwp
%          Options: .opt
%
% Options (opt structure):
%   .order         - Polynomial order: 0, 1, or 2 (default: 0)
%   .kernel_type   - Kernel: 'gaussian' or 'epanechnikov' (default: 'gaussian')
%   .y_type_out    - Output type: 'mean' or 'variance' (default: 'mean')
%   .N             - Grid size (default: auto based on sample size)
%   .accuracy      - NUFFT accuracy in digits (default: auto)
%   .dstd          - Number of standard errors for 1-SE bandwidth selection rule (default: 0)
%                    The 1-SE rule selects the largest bandwidth h such that:
%                    GCV(h) <= GCV_min + dstd * SE(GCV_min)
%                    dstd=0: minimum GCV; dstd=1: standard 1-SE rule
%                    Set to 5-20 for variance estimation to prefer smoother fits
%
% Algorithm:
%   1. Process complex-valued data (split into real/imaginary)
%   2. Set default options
%   3. Standardize data (z-score)
%   4. Create evaluation grid
%   5. Compute kernel density function in Fourier domain
%   6. Store all parameters in regs structure
%
% Author: Ying Wang, Min Li
% Create Time: 2021
% Copyright (c): 2020-2025 Ying Wang, yingwangrigel@gmail.com,
%                Min Li, minli.231314@gmail.com
% Joint China-Cuba LAB, UESTC
% License: GNU General Public License v3.0 (see LICENSE file)

%% Input validation
if nargin<4||isempty(opt)
    opt=[];
end
if nargin<3||isempty(h)
    h=[];
end

%% Process complex-valued predictors
% Complex predictors are split into real and imaginary parts
% Example: x = [1+2i, 3+4i] becomes x = [1, 3, 2, 4]
x_iscomplex=iscomplex(x,1);  % Check which columns are complex
x=[real(x),imag(x(:,x_iscomplex))];  % Split complex into [real, imag]

%% Get dimensions and validate
[Tx,dx]=size(x);  % Tx = number of samples, dx = number of predictors
if ismatrix(y)
    % Standard case: y is Ty x dy matrix
    [Ty,dy]=size(y);
    opt.y_corr_bandwidth=false;
elseif ndims(y)==3
    % Special case: different bandwidth for each response
    % y is Ty x dh x dy array
    [Ty,dhy,dy]=size(y);
    opt.y_corr_bandwidth=true;
end
assert(Tx==Ty,'fastLPR:create:DimensionMismatch', ...
    'Number of samples in x (%d) must equal number of samples in y (%d)', Tx, Ty);


%% Set default options

% Output type: 'mean' for E[Y|X] or 'variance' for Var[Y|X]
opt=set_defaults(opt,'y_type_out','mean');

% Kernel type: 'gaussian' (recommended) or 'epanechnikov'
opt=set_defaults(opt,'kernel_type','gaussian');

% Polynomial order: 0 (Nadaraya-Watson), 1 (local linear), 2 (local quadratic)
opt=set_defaults(opt,'order',0);

% Grid size N: number of evaluation points per dimension
% Rule of thumb: N ~= Tx^(1/dx) gives good balance of accuracy and speed
% Smaller N is faster but less accurate; larger N is more accurate but slower
opt=set_defaults(opt,'Nratio',1);  % Multiplier for default grid size
opt=set_defaults(opt,'N',ceil(opt.Nratio*(Tx^(1/dx))));
N=opt.N;

% Ensure N is a vector (broadcast scalar to all dimensions)
if isscalar(N)
    N = N * ones(1, dx);
end
opt.N = N;  % Update opt.N to be a vector
opt.Nratio=N(1)/(Tx^(1/dx));

% DEBUG: Print grid size
if isfield(opt, 'verbose') && opt.verbose
    fprintf('[DEBUG fastlpr_create] N = %d (grid size)\n', N);
end

% FFT padding: use power-of-2 length for speed
opt=set_defaults(opt,'flag_power2',true);

% NUFFT accuracy: number of accurate digits
% Auto-adjust based on sample size: fewer samples need higher accuracy
opt=set_defaults(opt,'accuracy',max(6-ceil(log10(Tx)),4));

% DEBUG: Print accuracy parameter
if isfield(opt, 'verbose') && opt.verbose
    fprintf('[DEBUG fastlpr_create] accuracy = %d (Tx=%d, N will be set next)\n', opt.accuracy, Tx);
end

% NUFFT deconvolution: apply correction for spreading kernel
% Only needed for very high accuracy (>6 digits)
opt=set_defaults(opt,'nufft_deconv',opt.accuracy>6);

% Bandwidth selection and degrees of freedom estimation
opt=set_defaults(opt,'dstd',0);  % Number of std for 1-SE rule (0 = minimum GCV, no 1-SE rule)
opt=set_defaults(opt,'calc_dof',true);  % Whether to calculate DOF
opt=set_defaults(opt,'num_dof_sample',10);  % Number of random samples for DOF estimation
opt=set_defaults(opt,'dof_seed',42);  % Random seed for DOF estimation (for reproducibility)


if ~isfield(opt,'y_grid_opt')
    % 'linear','cubic' or 'spline' et al. see griddedInterpolant
    opt.y_grid_opt.Method='linear';
    opt.y_grid_opt.ExtrapolationMethod='linear';
end

opt=set_defaults(opt,'compact',true);

% opt=set_defaults(opt,'x_iscomplex',iscomplex(x,1));
% opt=set_defaults(opt,'select_h',[]);
opt=set_defaults(opt,'y_corr_bandwidth',false);
opt=set_defaults(opt,'y_gcv_metric',[]);
opt=set_defaults(opt,'y_grid_method','griddedInterpolant');
opt=set_defaults(opt,'isgpu',false);


% opt=set_defaults(opt,'y_keep_all',false);
% opt=set_defaults(opt,'inherit_bandwidth',[]);



% opt=set_defaults(opt,'nufft_loop',true);



%% Initialize data types and precision
dtype=class(y);  % Store original data type

% Convert to appropriate precision based on accuracy requirement
% Higher accuracy needs higher precision (double), lower accuracy can use single
[x,y]=bit_convert(opt.accuracy,x,y);

% Handle GPU arrays if needed
if strcmp(dtype,'gpuArray') %#ok<STISA>
    x=gpuArray(x);
    y=gpuArray(y);
end
dtype=class(y);  % Update dtype after conversion

% Create colon indexing for multi-dimensional arrays
ndcolon(1:dx) = {':'};  % Example: for dx=2, ndcolon = {':', ':'}

%% Compute padded length L
if opt.flag_power2
    % Power-of-2 padding improves FFT speed and accuracy
    % Use L = 2^nextpow2(2*N-1) to avoid circular convolution
    % For multidimensional case, apply element-wise
    L = 2.^nextpow2(2*N-1);
else
    % Minimal padding (for testing or special cases)
    L=N+2;
end

% Ensure L is a vector (should already be from above, but double-check)
if isscalar(L)
    L = L * ones(1, dx);
end

%% Standardize data (z-score normalization)
% This ensures bandwidth h has consistent meaning across different scales
% After z-scoring: mean(x) ~= 0, std(x) = 1

xraw=x;  % Save original data
[x,~,x_std]=zscore(x);  % Standardize: x_new = (x - mean(x)) / std(x)

% Get data range after standardization
x_min=min(x);
x_max=max(x);
x_scale=x_max-x_min;  % Range of standardized data
x_delta=x_scale./N;   % Grid spacing

% Center data at origin (for symmetric padding)
x=x-(x_max+x_min)/2;
x_min=min(x);
x_max=max(x);
x_mean=mean(x);

%% Compute padding indices
% qin stores the start and end indices of the evaluation grid within the padded array
% This ensures the evaluation grid is centered in the padded array
qin=zeros(2,dx);

% Use symmetric padding for all N (especially odd N)
% Old method: round(L/2 - N/2) causes asymmetry for odd N
% New method: floor((L-N)/2) ensures symmetric padding
% Example: N=101, L=256 -> pad_left=77, pad_right=78 (symmetric)
pad_left = floor((L - N) / 2);
qin(1,:) = pad_left + 1;      % Start index of evaluation grid
qin(2,:) = pad_left + N;      % End index of evaluation grid


%% Compute kernel density function in Fourier domain
% This is the key step that enables O(N + M log M) complexity
% The kernel is precomputed in Fourier domain for fast convolution
[kdf,ihbad,dh,dh_input,h,lwp]=fastlpr_kdf(x,h,N,opt);

% Validate bandwidth dimensions for multi-response case
if opt.y_corr_bandwidth && dh~=dhy && dhy~=1
    error('fastLPR:create:BandwidthMismatch', ...
        'Number of bandwidths (%d) must match number of responses (%d)', dh, dhy);
end

%% Compute NUFFT knot positions
% Knots are the positions of data points in the frequency domain
% They must be in the range [-0.5, 0.5) for NUFFT

% Create index range for frequency domain (per dimension)
if dx == 1
    % 1D case: simple vectors
    idx_knot = (qin(1):qin(2))';  % Indices of evaluation grid
    idx_range = linspace(-1/2, 1/2-1/L, L)';  % Frequency range [-0.5, 0.5)
    hf = 1/L;

    % Map data points to frequency domain
    if N ~= L
        % With padding: map to evaluation grid range
        knot_scale = idx_range(idx_knot(end)) - idx_range(idx_knot(1));
        knot = (x - x_min) ./ x_scale .* knot_scale + idx_range(idx_knot(1));
    else
        % No padding: map to full range
        knot_scale = idx_range(end) - idx_range(1);
        knot = (x - x_min) ./ x_scale .* knot_scale - 0.5;
    end
else
    % Multidimensional case: cell arrays for each dimension
    idx_knot = cell(1, dx);
    idx_range = cell(1, dx);
    hf = zeros(1, dx);
    knot = zeros(size(x));
    knot_scale = zeros(1, dx);  % Initialize knot_scale for multi-D

    for d = 1:dx
        idx_knot{d} = (qin(1,d):qin(2,d))';
        idx_range{d} = linspace(-1/2, 1/2-1/L(d), L(d))';
        hf(d) = 1/L(d);

        % Map data points to frequency domain for this dimension
        if N(d) ~= L(d)
            % With padding: map to evaluation grid range
            knot_scale(d) = idx_range{d}(idx_knot{d}(end)) - idx_range{d}(idx_knot{d}(1));
            knot(:,d) = (x(:,d) - x_min(d)) ./ x_scale(d) .* knot_scale(d) + idx_range{d}(idx_knot{d}(1));
        else
            % No padding: map to full range
            knot_scale(d) = idx_range{d}(end) - idx_range{d}(1);
            knot(:,d) = (x(:,d) - x_min(d)) ./ x_scale(d) .* knot_scale(d) - 0.5;
        end
    end
end

% Compute NUFFT spreading width based on accuracy requirement
% CRITICAL FIX: Pass L (padded grid size), not Ty (sample size)!
% The NUFFT grid parameters should be based on the FFT grid size, not the number of data points
[Msp, Mr_check, tau_check, ratio_check] = compute_grid_params(L, opt.accuracy);

% DEBUG: Print grid parameters
if isfield(opt, 'verbose') && opt.verbose
    fprintf('[DEBUG fastlpr_create] Grid params: L=%d, Msp=%d, Mr=%d, tau=%.2e, ratio=%d\n', ...
        L, Msp, Mr_check, tau_check, ratio_check);
end

%% Compute output extraction indices
% qout stores the indices to extract the evaluation grid from the padded result
if dx == 1
    qout = (L - N) / 2;
    qout = [ceil(qout) + 1, ceil(qout) + N];
else
    qout = (L - N) / 2;  % Element-wise for vectors
    qout = [ceil(qout)' + 1, ceil(qout)' + N'];  % dx*2 matrix
end

%% Create evaluation grid in original (raw) space
% This grid is used for output and for parametric interpolation methods
xlist = multispace(min(xraw), max(xraw), N);  % Create N-point grid for each dimension

% multispace returns:
% - For 1D (dx=1): (1 x N) matrix
% - For multi-D with same N: (dx x N) matrix
% - For multi-D with different N: {dx x 1} cell array

if dx == 1
    % 1D case: convert to cell
    xlist = {xlist'};  % Transpose to column vector and wrap in cell
elseif ~iscell(xlist)
    % Multi-D with same N: convert matrix (dx x N) to cell array {dx x 1}
    % Each cell contains one row (1 x N) transposed to (N x 1)
    xlist_cell = cell(dx, 1);
    for d = 1:dx
        xlist_cell{d} = xlist(d, :)';  % Extract row d and transpose
    end
    xlist = xlist_cell;
else
    % Multi-D with different N: already cell array, just transpose each
    for d = 1:dx
        xlist{d} = xlist{d}';
    end
end


%% Store all parameters in regression structure
% The regs structure contains all information needed for fast LPR

% Data
regs.xraw=xraw;      % Original predictor data (before standardization)
regs.x=x;            % Standardized predictor data
regs.y=y;            % Response data
regs.Tx=Tx;          % Number of samples
regs.Ty=Ty;          % Number of response samples
regs.dx=dx;          % Number of predictor dimensions
regs.dy=dy;          % Number of response variables
regs.dtype=dtype;    % Data type (single, double, gpuArray)
regs.Msp=Msp;        % NUFFT spreading width

% Grid parameters
regs.L=L;            % Padded length (for FFT)
regs.N=N;            % Grid size (evaluation points per dimension)
regs.ndcolon=ndcolon;% Colon indexing for multi-dimensional arrays

% Standardization parameters
regs.x_min=x_min;    % Minimum of standardized data
regs.x_max=x_max;    % Maximum of standardized data
regs.x_scale=x_scale;% Range of standardized data
regs.x_mean=x_mean;  % Mean of standardized data
regs.x_std=x_std;    % Standard deviation (from zscore)
regs.x_delta=x_delta;% Grid spacing

% Kernel parameters
regs.h=h;            % Bandwidth (dh x dx matrix)
regs.kdf=kdf;        % Kernel density function in Fourier domain
regs.ihbad=ihbad;    % Logical array indicating bad bandwidths
regs.dh=dh;          % Number of valid bandwidths
regs.dh_input=dh_input;  % Original number of bandwidths

% Index parameters
regs.qin=qin;        % Padding indices (start and end of evaluation grid)
regs.qout=qout;      % Output extraction indices

% NUFFT parameters
regs.knot_scale=knot_scale;  % Scaling factor for knot positions
regs.knot=knot;      % NUFFT knot positions (data points in frequency domain)
regs.hf=hf;          % Frequency spacing

% Output grid
regs.xlist=xlist;    % Evaluation grid in original space (cell array)

% Options
regs.opt=opt;        % All options

% Response data properties
regs.y_isreal=~iscomplex(y,1);  % Logical array indicating real-valued responses

% Local polynomial parameters
regs.lwp=lwp;        % Local polynomial estimator functions (for order >= 1)

end